File: app/models/distribucion_operativa.py
from app.extensions import db


class DistribucionOperativa(
    db.Model
):

    __tablename__ = (
        "distribucion_operativa"
    )

    id = db.Column(
        db.Integer,
        primary_key=True
    )

    fecha = db.Column(
        db.Date,
        nullable=False
    )

    contrato = db.Column(
        db.Text,
        nullable=False
    )

    # Sede donde opera esta cuadrilla (informativo; la sede oficial viene del Contrato)
    sede = db.Column(db.String(150), nullable=True)

    recurso = db.Column(
        db.String(200),
        nullable=False
    )

    placa = db.Column(
        db.String(50)
    )

    orden_trabajo = db.Column(
        db.String(100)
    )

    tipo_actividad = db.Column(
        db.Text
    )

    tipo_cuadrilla = db.Column(
        db.Text
    )

    hora_salida_sede = db.Column(
        db.Time
    )

    hora_llegada_sede = db.Column(
        db.Time
    )

    cedula_1 = db.Column(
        db.String(50)
    )

    cedula_2 = db.Column(
        db.String(50)
    )

    cedula_3 = db.Column(
        db.String(50)
    )

    cedula_4 = db.Column(
        db.String(50)
    )

    cedula_5 = db.Column(
        db.String(50)
    )

    numero_celular = db.Column(
        db.String(50)
    )

    latitud = db.Column(
        db.String(100)
    )

    longitud = db.Column(
        db.String(100)
    )

    duracion_actividad = db.Column(
        db.String(50)
    )

    observacion = db.Column(
        db.Text
    )

    meta = db.Column(
        db.Float
    )

File: app/static/js/distribucion_operativa.js
// Distribución Operativa — solo lectura, datos desde GPS Monitor
// ====================================================

function fmtFecha(val) {
    if (!val) return "";
    const p = String(val).split("-");
    return p.length === 3 ? `${p[2]}/${p[1]}/${p[0]}` : val;
}

function origenBadge(cell) {
    const v = (cell.getValue() || "").toLowerCase();
    if (v === "gps_monitor") {
        return `<span class="badge" style="background:#0ea5e9;font-size:11px;">GPS Monitor</span>`;
    }
    return `<span class="badge bg-secondary" style="font-size:11px;">Manual</span>`;
}

let tabla = new Tabulator("#tablaDistribucion", {
    data: datosDistribucion,
    layout: "fitDataStretch",
    height: "72vh",
    pagination: true,
    paginationSize: 50,
    resizableColumns: true,
    movableColumns: true,
    clipboard: true,
    clipboardCopyStyled: false,
    clipboardCopyConfig: { rowHeaders: false, columnHeaders: true },
    columns: [
        {
            title: "Fecha",
            field: "fecha",
            frozen: true,
            formatter: function(cell) { return fmtFecha(cell.getValue()); },
            headerFilter: true,
            headerFilterPlaceholder: "DD/MM/AAAA",
            headerFilterFunc: function(hv, rv) {
                if (!hv) return true;
                const raw = String(rv || "");
                if (raw.includes(hv)) return true;
                const p = raw.split("-");
                return p.length === 3 ? `${p[2]}/${p[1]}/${p[0]}`.includes(hv) : false;
            },
            width: 110,
        },
        {
            title: "Contrato",
            field: "contrato",
            headerFilter: true,
            minWidth: 220,
        },
        {
            title: "Recurso",
            field: "recurso",
            headerFilter: true,
            minWidth: 140,
        },
        {
            title: "Placa",
            field: "placa",
            headerFilter: true,
            width: 100,
        },
        {
            title: "Orden Trabajo",
            field: "orden_trabajo",
            headerFilter: true,
            minWidth: 130,
        },
        {
            title: "Tipo Actividad",
            field: "tipo_actividad",
            headerFilter: true,
            minWidth: 160,
        },
        {
            title: "Tipo Cuadrilla",
            field: "tipo_cuadrilla",
            headerFilter: true,
            minWidth: 150,
        },
        {
            title: "Cédula 1",
            field: "cedula_1",
            headerFilter: true,
            width: 110,
        },
        {
            title: "Nombre",
            field: "nombre_1",
            headerFilter: true,
            minWidth: 180,
        },
        {
            title: "Cédula 2",
            field: "cedula_2",
            headerFilter: true,
            width: 100,
        },
        {
            title: "Cédula 3",
            field: "cedula_3",
            headerFilter: true,
            width: 100,
        },
        {
            title: "Cédula 4",
            field: "cedula_4",
            headerFilter: true,
            width: 100,
        },
        {
            title: "Cédula 5",
            field: "cedula_5",
            headerFilter: true,
            width: 100,
        },
        {
            title: "Meta",
            field: "meta",
            headerFilter: false,
            width: 120,
            hozAlign: "right",
            formatter: function(cell) {
                const v = cell.getValue();
                if (v === null || v === undefined || v === "") return "—";
                return "$ " + Number(v).toLocaleString("es-CO", {
                    minimumFractionDigits: 0,
                    maximumFractionDigits: 0
                });
            },
        },
        {
            title: "Duración (min)",
            field: "duracion_actividad",
            headerFilter: true,
            width: 120,
            hozAlign: "right",
        },
        {
            title: "Latitud",
            field: "latitud",
            headerFilter: true,
            width: 120,
            formatter: function(cell) {
                const v = cell.getValue();
                return v ? String(v).replace(".", ",") : "";
            },
        },
        {
            title: "Longitud",
            field: "longitud",
            headerFilter: true,
            width: 120,
            formatter: function(cell) {
                const v = cell.getValue();
                return v ? String(v).replace(".", ",") : "";
            },
        },
        {
            title: "Observación",
            field: "observacion",
            headerFilter: true,
            minWidth: 200,
        },
        {
            title: "Origen",
            field: "origen",
            formatter: origenBadge,
            headerFilter: true,
            width: 120,
        },
    ],
});

// Exportar Excel
const btnExportar = document.getElementById("btnExportar");
if (btnExportar) {
    btnExportar.addEventListener("click", function() {
        tabla.download("xlsx", "distribucion_operativa.xlsx", { sheetName: "Distribución" });
    });
}

// Actualizar contador al filtrar
tabla.on("dataFiltered", function(filters, rows) {
    const el = document.getElementById("totalRegistros");
    if (el) el.textContent = rows.length;
});

// El filtro de contrato se maneja desde el template (acceso a variable tabla via window)
